tsdb/head.go

```diff
@@ -571,31 +620,106 @@ func (h *Head) Init(minValidTime int64) error {
 		return nil
 	}
 
-	level.Info(h.logger).Log("msg", "replaying WAL, this may take awhile")
-	// Backfill the checkpoint first if it exists.
-	dir, startFrom, err := wal.LastCheckpoint(h.wal.Dir())
-	if err != nil && err != record.ErrNotFound {
-		return errors.Wrap(err, "find last checkpoint")
-	}
+	refSeries := map[uint64]*memSeries{}
+	level.Info(h.logger).Log("msg", "creating series from WAL")
 	multiRef := map[uint64]uint64{}
-	if err == nil {
-		sr, err := wal.NewSegmentsReader(dir)
-		if err != nil {
-			return errors.Wrap(err, "open checkpoint")
+	// Backfill series from the checkpoint first if it exists.
+	if err := h.executeOnCheckpointReader(func(checkpointReader *wal.Reader) error {
+		// A corrupted checkpoint is a hard error for now and requires user
+		// intervention. There's likely little data that can be recovered anyway.
+		if err := h.loadSeries(checkpointReader, multiRef, refSeries); err != nil {
+			return errors.Wrap(err, "load series from checkpoint")
 		}
-		defer func() {
-			if err := sr.Close(); err != nil {
-				level.Warn(h.logger).Log("msg", "error while closing the wal segments reader", "err", err)
+		return nil
+	}); err != nil {
+		return err
+	}
+	// Backfill series from the remaining WAL.
+	if err := h.executeOnWALSegmentReader(func(segmentReader *wal.Reader) error {
+		if err := h.loadSeries(segmentReader, multiRef, refSeries); err != nil {
+			return errors.Wrap(err, "load series from WAL")
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	unknownRefs := 0
+	if err := h.chunkReadWriter.IterateAllChunks(func(seriesRef, chunkRef uint64, mint, maxt int64) error {
+		ms := refSeries[seriesRef]
+		if ms == nil {
+			unknownRefs++
+			return nil
+		}
+
+		if len(ms.mmappedChunks) > 0 {
+			if ms.mmappedChunks[len(ms.mmappedChunks)-1].maxTime >= mint {
+				return errors.Errorf("out of sequence m-mapped chunk for series %s", ms.lset.String())
 			}
-		}()
+		}
+
+		ms.mmappedChunks = append(ms.mmappedChunks, &mmappedChunk{
+			ref:     chunkRef,
+			minTime: mint,
+			maxTime: maxt,
+		})
+		return nil
+	}); err != nil {
+		return errors.Wrap(err, "iterate on-disk chunks")
+	}
 
+	level.Info(h.logger).Log("msg", "replaying WAL, this may take awhile")
```

Comment:
hm, we do lot's of code even before that logline. Should we move it to the beginning of this method?

Also can we add `finished replaying WAL` with the extra field `elapsed` that measures elapsed time? It's always useful for long-running task (: